### ISSUE TITLE
write functions that work work when numpy is not installed

### ISSUE THREAD

**Issue by santoshphilip** [santoshphilip/eppy#30]
There are geometry functions in ./eppy/geometry that use numpy.
It can be difficult to install numpy at times (for instance if you want to run epp in rhino <http://www.rhino3d.com>)

- rewrite the those functions to run without numpy. 
- test if numpy is installed (use <try .. except> block to check)
- if it is not installed, use the new functions

**Comment by eayoungs** [comment]
Santosh

I've revised the vector cross product function and tests to use exception handling and assert that exceptions are raised in the tests. I just got the tests to pass and will be moving on to use the same pattern to revise the rest of the functions I've written so far (dot product, determinant) before moving on to other functions.

Please review and advise as to weather the current approach is satisfactory before I move on

Thx!

Eric

**Comment by santoshphilip** [comment]
Awesome !
I'll review and respond. 

Santosh.

**Comment by santoshphilip** [comment]
Eric, 
    You are on the right track. Please continue.
    
I have some thoughts on coding style (PEP 8)
But please do not do anything about code style right now.
There automized tools to review the style and update. 
Once you are finished we will apply those tools.

Looks like you are close to wrapping this up.

Santosh

**Comment by eayoungs** [comment]
Great! Thanks for the feedback. I feel like I'm starting to get the hang of it and hope to contribute more.

**Comment by santoshphilip** [comment]
Eric,    You have been programming rather defensively so far, trying to get things right.You are on the right track.You can now code a little more aggressively now.
In fact that is the thinking behind unit testing.- you write the test first- the test will fail because you have no code- write the code- the test may succeed or fail- write more tests and continue- stop when you think it is good enough - not when it is perfect.
Basically you making the computer catch your mistakes.
Later you clean the style, refactor, document better etc.
At the end you transition to python3.

Santosh

**Comment by eayoungs** [comment]
Sounds good. I will keep that in mind and try to follow your advice. Sounds like the idea is to move more quickly and allow for more experimentation. I'm excited to start a new phase in my coding!

**Comment by eayoungs** [comment]
Santosh,

I started writing the replacement to numpy.array(), which encompasses a lot of functionality. I was thinking I might need to write a class since there are many methods associated with it.

Then I took a look at where it's actually used within eppy. I only find references to it in the *int2lines* module. Furthermore, I don't see it imported or even referenced anywhere else in the project. If this is the case, I wonder if there is a good justification for re-implementing it in pure Python. I will move on to the last 2 numpy functions used; square root, and arc-cosine.

Cheers

Eric

**Comment by santoshphilip** [comment]
don't write anything you don't have to
Also see if you can find the other function in the math library
do 
import math 
and check

**Comment by vinc85** [comment]
How can I import eppy in a java web application using Jython, with dependencies of numpy? thanks

**Comment by santoshphilip** [comment]
Eric,
    To get a background in this, take a look at issue #48 
and my comment in
https://groups.yahoo.com/neo/groups/eppy/conversations/messages/3

Vinc85 can use eppy in jython once you remove the numpy dependency.
Essentially once you are done with this and close this issue.

Vinc85 is wondering when that will happen :-)
Seems to me that you are very close to being done

**Comment by eayoungs** [comment]
Yes, I was just looking into this. I was able to replace the remaining numpy dependencies with the Python native library, "math" so essentially I am done. I was working on a custom exception for singular matrices. Right now I've compromised and simply manually raised the base Exception class when the determinant is zero.
I wonder if I/we should do some integration testing?
vinc85 if you want to use the code, it should all be available in the most recent commit of the branch: "i30_pythongeometry"

vinc85: Great timing! This is quite validating and makes my work seem very useful. It's quite gratifying.

Cheers

Eric

**Comment by santoshphilip** [comment]
If it passes all the unit tests without installing numpy, you are done

**Comment by eayoungs** [comment]
Cool! I will set up a virtual environment and run the tests, then.

**Comment by vinc85** [comment]
Many thanks Eric, I have tested eppy "i30_pythongeometry" and it seems to work without the library numpy.
I hope it works now importing eppy in my java web application in Netbeans. I urge to run eppy on Web server.

 


     Il Martedì 16 Giugno 2015 1:27, Eric Youngson <notifications@github.com> ha scritto:
   

 Cool! I will set up a virtual environment and run the tests, then.—

**Comment by santoshphilip** [comment]
Also test if all the code in ./docs/Main_Tutorial.ipynb work.
Open it up in ipython notebook and run all the cells.

**Comment by vinc85** [comment]
Dear all,I have imported a new version of eppy (without numpy dependecies) in a Java Web Application and when I doing this
iddfile = request.getContextPath() + '/Energy+V7_2_0.idd'fname1 = request.getContextPath() + '/smallfile.idf'IDF.setiddname(iddfile)idf1 = IDF(fname1)
I have this error:
Traceback (most recent call last):
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/serv.py", line 18, in doPost
    idf1 = IDF(fname1)
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/modeleditor.py", line 615, in __init__
    super(IDF3, self).__init__(idfname)
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/modeleditor.py", line 560, in __init__
    super(IDF2, self).__init__(idfname)
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/modeleditor.py", line 488, in __init__
    super(IDF1, self).__init__(idfname)
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/modeleditor.py", line 442, in __init__
    self.read()
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/modeleditor.py", line 472, in read
    readout = idfreader1(self.idfname, self.iddname,
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/idfreader.py", line 148, in idfreader1
    versiontuple = iddversiontuple(iddfile)
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/idfreader.py", line 35, in iddversiontuple
    line1 = fhandle.readline()
AttributeError: 'unicode' object has no attribute 'readline'
 Why?
Many thanks.


     Il Mercoledì 17 Giugno 2015 1:44, santoshphilip <notifications@github.com> ha scritto:
   

 Also test if all the code in ./docs/Main_Tutorial.ipynb work.
Open it up in ipython notebook and run all the cells.—

**Comment by santoshphilip** [comment]
vinc85,
I have reformated your reply to make it more readable.
see 
https://guides.github.com/features/mastering-markdown/
on how to do this


### vince85's reply reformatted


Dear all,I have imported a new version of eppy (without numpy dependecies) in a Java Web Application and when I doing this

    iddfile = request.getContextPath() + '/Energy+V7_2_0.idd'fname1 = request.getContextPath() + '/smallfile.idf'
    IDF.setiddname(iddfile)
    idf1 = IDF(fname1)

I have this error:
```
Traceback (most recent call last):
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/serv.py", line 18, in doPost
    idf1 = IDF(fname1)
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/modeleditor.py", line 615, in __init__
    super(IDF3, self).__init__(idfname)
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/modeleditor.py", line 560, in __init__
    super(IDF2, self).__init__(idfname)
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/modeleditor.py", line 488, in __init__
    super(IDF1, self).__init__(idfname)
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/modeleditor.py", line 442, in __init__
    self.read()
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/modeleditor.py", line 472, in read
    readout = idfreader1(self.idfname, self.iddname,
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/idfreader.py", line 148, in idfreader1
    versiontuple = iddversiontuple(iddfile)
  File "/Users/vincenzo/NetBeansProjects/ProvaPython/build/web/eppy/idfreader.py", line 35, in iddversiontuple
    line1 = fhandle.readline()
AttributeError: 'unicode' object has no attribute 'readline'
```

**Comment by santoshphilip** [comment]
vinc85,
    That looks like a bug to me. I'll try to fix it in this branch

**Comment by vinc85** [comment]
Ok, but I have not this error if I run eppy on desktop and not on a Web Application?Why?
Many thanks. 


     Il Mercoledì 17 Giugno 2015 16:48, santoshphilip <notifications@github.com> ha scritto:
   

 vinc85,
 That looks like a bug to me. I'll try to fix it in this branch—

**Comment by santoshphilip** [comment]
It works if the filename is a string
Does not work if the filename is in unicode
see here about string and unicode:
https://docs.python.org/2/howto/unicode.html

java is sending the filename in unicode, while the desktop script is sending it in string

eppy should work with unicode too. 
This needs to be fixed.

**Comment by vinc85** [comment]
It's true!!! ;-)
 


     Il Mercoledì 17 Giugno 2015 17:06, santoshphilip <notifications@github.com> ha scritto:
   

 It works if the filename is a string
Does not work if the filename is in unicode
see here about string and unicode:
https://docs.python.org/2/howto/unicode.htmljava is sending the filename in unicode, while the desktop script is sending it in stringeppy should work with unicode too. 
This needs to be fixed.—

**Comment by santoshphilip** [comment]
vince85,  try it now. 

eayoungs (Eric),  unit test is failing if I run it with numpy installed.
Can you look into this ?

Santosh

**Comment by vinc85** [comment]
I don't understand...
 


     Il Mercoledì 17 Giugno 2015 17:52, santoshphilip <notifications@github.com> ha scritto:
   

 vince85, try it now. eayoungs (Eric), unit test is failing if I run it with numpy installed.
Can you look into this ?Santosh—

**Comment by santoshphilip** [comment]
I fixed the bug. 
Download and try again. 
Your code should work. 

My other comment was directed to Eric. 
I does not affect you. 

Santosh.

**Comment by vinc85** [comment]
Ok, many thanks ;-)

Vincenzo

**Comment by eayoungs** [comment]
Sure Santosh; I will take a look. I was having some trouble with my virtual environment yesterday but I should have it figured out soon.

**Comment by eayoungs** [comment]
Santosh,
All the tests for the branch seem to work with and without numpy installed on my machine (see below for details). Can you tell me more about which unit test fails and which packages are installed?

Thx!

-------------------------------------------------------------------------------------------------------------------------------------------
(MacBook Pro 2014, OSX Yosemite / Python 2.7.10, virtualenv 13.03)
$ *pip list*:
beautifulsoup4 (4.3.2)
bunch (1.0.1)
eppy (0.4.6.4a0, /Users/eayoungs/repo/Code/SimTools/eppy)
pip (7.0.3)
py (1.4.28)
pydot (1.0.2)
pyparsing (1.5.7)
pytest (2.7.1)
setuptools (17.0)
wheel (0.24.0)

**Comment by santoshphilip** [comment]
I'll double check on this.
I assume we are speaking of branch "i30_pythongeometry"

**Comment by santoshphilip** [comment]
with numpy installed and running py.test I am getting 
```
    def angle2vecs(vec1,vec2):
        # vector a * vector b = |a|*|b|* cos(angle between vector a and vector b)
        dot = np.dot(vec1,vec2)
        vec1_modulus = math.sqrt((vec1*vec1).sum())
        vec2_modulus = math.sqrt((vec2*vec2).sum())
        if (vec1_modulus * vec2_modulus) == 0:
            cos_angle = 1
        else: cos_angle = dot / (vec1_modulus * vec2_modulus)
>       return math.degrees(math.arccos(cos_angle))
E       AttributeError: 'module' object has no attribute 'arccos'

eppy/geometry/surface.py:95: AttributeError

```

**Comment by santoshphilip** [comment]
I am on branch i30_pythongeometry
commit is bf7c05b60c6d1dbf97bc1a3e9ed754f5a99de160 
This is the latest commit on that branch .
It gave the above error

I also tested commit 1cc64fa0702340c092d115bd7b03725008523156 , the last commit you made on that branch. It gave the same error

Are we both testing the same commit ?

**Comment by santoshphilip** [comment]
pip install yolk

in your virtualenv. 

if you type 
    
    yolk -l

it will list all the packages installed in that virtualenv

**Comment by santoshphilip** [comment]
I think numpy is not installed in your test where you think it is installed.
You can also check this running the following from the shell 

    python -c 'import numpy'

if numpy is not installed, it will throw an ImportError

**Comment by eayoungs** [comment]
Mmm, yes. This is an oversight on my part. It should be math.*acos*. I fix this and check for any more issues.

**Comment by eayoungs** [comment]
I simply used a try/except pattern to import either numpy or, if not available, py_numeric to substitute but, on closer inspection, it looks like I might have to come up with a better way to replace numpy function calls within surface.py since the determinant function is stored another level down in the linalg module: *np.linalg.det()*

Also with the functions available from the math and array libraries, I simply renamed the function calls that are available in the math.py library but I realize that this may not be the best strategy if numpy is faster, then it should be used when available.

I'm wondering if we will need to have two different modules for each case and switch them when called elsewhere in the program. This doesn't seem ideal, though.

Any thoughts on how to abstract the function calls so within the module?

**Comment by santoshphilip** [comment]
Eric,
    Right now you are doing:

```
try:
    import numpy as np
except ImportError as e:
    import py_numeric as np
```

would the following solve the problem ?

```
try:
    import numpy as np
    import numpy.arccos as arccos # do same for other functions you need
except ImportError as e:
    import py_numeric as np
    import math.acos as arccos # do same for other functions you need
```

**Comment by eayoungs** [comment]
This does help. Although the issue still remains with the *linalg* submodule. I've moved the determinant function into a separate module to mimic the structure and put the import statement into the try/except pattern as you described above and it seems to work as far as that goes, but I've discovered a new issue.
I designed the determinant function to take 3 arguments as vectors but it looks like numpy takes a list of lists containing the vectors as a single parameter. It should be strait-forward to rewrite and parse the incoming list.

**Comment by eayoungs** [comment]
On another note, I found an interesting library that may help with limiting numpy dependecies,   
    https://github.com/wadetb/tinynumpy
I think it only replaces the ndarray class but it may come in handy.

**Comment by eayoungs** [comment]
I'm running into a type error with a particular unit test:
eppy/tests/geometry_tests/test_surface.py - test_tilt() > tilt(poly) > unit_normal() > np.linalg.det()

The test returns 0, correctly but the numpy.float64 type passes with a warning while all the types I can access produce a divide by zero error. I can continue to reproduce as much of the determinant function as possible but it starts to get complicated, quickly and the core function seems to be implemented in C.
I've pulled out the relevant code from the failing test and put it into an IPython notebook here
https://gist.github.com/eayoungs/1c60eb2eb09de812bd4f

**Comment by jamiebull1** [comment]
Ah, that's actually my code that Sanjay dug up on ActiveState before I'd even heard of Eppy. The solution is, in eppy.geometry.surface.py, to explicitly test for the case where the points are all in line. The easiest way to do this is to test after the for loop in `area(poly)`. It's also more efficient to test after the calculation rather than test for colinearity beforehand, given that it's probably rare that the function will be passed vertices that are in a straight line.

```python
def area(poly):
    """Area of a polygon poly""" 
    if len(poly) < 3: # not a plane - no area
        return 0
    total = [0, 0, 0]
    N = len(poly)
    for i in range(N):
        vi1 = poly[i]
        vi2 = poly[(i+1) % N]
        prod = np.cross(vi1, vi2)
        total[0] += prod[0]
        total[1] += prod[1]
        total[2] += prod[2]
    if total == [0, 0, 0]:  # points are in a straight line
        return 0
    result = np.dot(total, unit_normal(poly[0], poly[1], poly[2]))
    return abs(result/2)
```

**Comment by santoshphilip** [comment]
That is indeed Jamie's code I used without attribution (since I did not know he wrote it)
So in a sense Jamie was already a collaborator on eppy.
(and somehow I became Sanjay instead of Santosh :-) 

Eric,
     I am not paying too much attention to the specifics here. Will do so if you need me to. (all is well as long as unit tests pass at some point)

**Comment by jamiebull1** [comment]
Haha, sorry about that. Now I understand why people never notice when they call me Jaime!

**Comment by eayoungs** [comment]
Thanks so much! I'm glad I asked. :-)
Wait, now I'm confused; you're name isn't Jaimie?

No worries Santosh; I'm on task now.

**Comment by jamiebull1** [comment]
Jamie. But living in Spain, I've got quite used to Jaime. Jaimie is a new one though!

**Comment by eayoungs** [comment]
I've got all the tests passing now. I have a lot of cleaning up I'd like to do still but it seems to be working. I replaced the array class with a small open source project called [tinynumpy](https://github.com/wadetb/tinynumpy) and I'd like to implement all the functions I've created within that module for clarity and so I can contribute the code back to that project.

**Comment by santoshphilip** [comment]
have you pushed all changes back ?

with numpy **NOT** installed, I get the following error with pytest:

    ==================================== ERRORS ====================================
    ________ ERROR collecting eppy/tests/geometry_tests/test_volume_zone.py ________
    eppy/tests/geometry_tests/test_volume_zone.py:22: in <module>
        import eppy.geometry.volume_zone as volume_zone
    eppy/geometry/volume_zone.py:25: in <module>
        import numpy as np
    E   ImportError: No module named numpy

**Comment by eayoungs** [comment]
Yes, I think that is something to do with the error handling in the
try/except statement. Without that statement, they all pass; without numpy.
With numpy, they still all pass as well, but the exception seems to be
picked up by pytest.
It's not ready to be merged back to the master branch but the unit tests,
by themselves, are passing. It will likely ready for a merge by early next
week.

**Comment by eayoungs** [comment]
OK, now it looks like all tests are passing with and without numpy installed. Turns out there were some simple math functions (add, subtract, divide) that involved the ndarray type from numpy. I'm not sure how I overlooked that during the previous testing but everything seems to run and pass now, so go ahead and take another look.
I'm still planning on cleaning things up a bit this weekend before merging but it should be functional.

Cheers

Eric

**Comment by santoshphilip** [comment]
let me do the final merge into develop
Just let me know when.

Santosh

**Comment by eayoungs** [comment]
OK, Santosh; I think the branch is ready for a merge. Feel free to let me know if there are any changes you'd like me to make.

Cheers

Eric

**Comment by santoshphilip** [comment]
I'll do this merge after I do the release with the MIT license.
(trying to do one thing at t time :-)

**Comment by eayoungs** [comment]
Sure, no worries. It's nice to see so much activity.

**Comment by santoshphilip** [comment]
all tests are passing with 
- numpy and 
- without numpy.

On my work computer one of the tests is for tinynumpy is failing.
This is very curious.

I'll investigate further and report back

**Comment by eayoungs** [comment]
On my machine there are 2 tests that are skipped in tinynumpy. One is skipped both with and without numpy installed, the other is skipped when numpy is not installed.

**Comment by santoshphilip** [comment]
I did notice that.
Something on my work machine is different that is triggering the failure.
maybe an older version of pytest

I have succeeded in merging your branch into a copy of the develop branch.
It had a couple of merge conflicts that had to be hand edited.
The merged branch passes all tests

I'll now run pylint and clean up some stuff. 
(pylint sends to be very strict, and I tend to ignore some it's complaints)

Then I'll generate the python3 files and run tests in python3 too.

**Comment by santoshphilip** [comment]
i had not seen `skip` used in pytest before.
Learnt something new here

**Comment by santoshphilip** [comment]
Eric,
    
Can you check if tinynumpy works under python3. The github page claims that it does. I want to make sure since eppy should run under python3.

tinynumpy is not on pypi. 
It would be nice to be able to do 
- pip install tinynumpy 

and then not have to worry about it, especially if it will install either python2 or python3 as appropriate
In any case check into this and if things are not working we would expect, we'll find a workaround. Maybe we can persuade the developers to upload to pypi

**Comment by eayoungs** [comment]
Sure I will test it for 3 compatibility and ask [Wade](https://github.com/wadetb) if he can get the project set up on pypi. I was considering whether to try to import tinynumpy into eppy rather than (as I actually did) simply embed the modules inside eppy. I figured for right now it's faster to just to embed tinynumpy as it's a small project, early in it's lifecycle and so not changing very quickly but in the long term I'd like to move toward an externally imported module.
I've also committed back the functions I created to tinynumpy. This could be a useful relationship between projects.

Eric ToDo:
- [ ] Test tinynumpy under Python3
- [ ] Determine a path for tinynumpy to be available through pip

**Comment by eayoungs** [comment]
As for the tests, something to keep in mind is that tinynumpy's tests were written with unittest. I did make a simple change to get it to work with py.test (I don't remember exactly what it was).
Glad I wasn't the only one who hadn't heard of skip; I looked it up too.

**Comment by santoshphilip** [comment]
- if tinynumpy does **not** work under python3
    - the eppy team can try to make that happen. 
    - or
    - we can wait for Wade to do it.
        - in the meantime, eppy release for python3 will have numpy as a dependency.

In essence, tinynumpy not working under python3 is not a deal breaker. It does not stop our release cycle

**Comment by santoshphilip** [comment]
Eric.

I started to merge branch 'i30_pythongeometry' into a copy of develop. A number of additional steps had to be taken:
- fix some merge conflicts by hand. (most of them were done in the merge tool)
- add MIT license to the new files
- get files ready for python3 transition
- run pylint to come close to PEP 0008 style
- use 2to3 to transition to python3
- unit test for python3 code

A number of tools have to be set up in your development environment to do this.
My thinking on this is this:
- I complete all the steps to do the merge.  (into a copy of develop)
- I commit at all important stages, with a clear commit message
- If you look at the log, you can see the steps I have taken.
- You can also do a diff between the commits to see what has changed.
- Now you can do the merge on your own, by looking at my steps
- this may help to shorten the learning curve, and to set up the tools in your environment.
- some of this may go into the developer documentation to help others do this (Leora and myself will take responsibility for this)

Would you like me to do this ?
So you have any other ideas ?

**Comment by eayoungs** [comment]
You know, it just occurs to me that I've been working from the master branch rather than pulling changes from the develop branch. I wonder if this has caused any issues?

Whatever setup is required to do the job, I'm happy to set it up on my system. As far as PEP 8 is concerned, I was checking in for quite a while but recently stopped just to concentrate on getting the tests to pass.

I think setting up this branch for Python 3 would be especially useful for me, if I'm going to work on the release.

**Comment by santoshphilip** [comment]
Right now develop and master are identical.
In general
- branch from master and work on an issue
- merge into develop when you are done.
- I will merge develop into master before a release

Regarding PEP 8. Be lazy about it. Code in any way you want. Let pylint do the hard work of getting you to PEP 8

**Comment by eayoungs** [comment]
Santosh,
    All sounds good. Do you think it would be useful for me to start documenting the contribution process while it's still fresh in my mind? I'm thinking I could just add some markdown to the repo.

**Comment by santoshphilip** [comment]
It may be so.
Why don't you make some notes

**Comment by eayoungs** [comment]
Santosh,
    Tinynumpy does work with Python 3. I ran the tests with my additions and found a small change in my code needed for Python 3 compatibility. I've made the change in this issue branch (*i30_pythongeometry*) and pushed the branch back to the repo.
    Also, I pulled all changes (quite a few) from the master branch into *i30_pythongeometry* and made a few merges by hand. I hope this will make the merging into the development branch easier.

Hope this helps!

**Comment by santoshphilip** [comment]
Done …. Your merges made things easier.
I have merged it into develop. I did the following things
- Put MIT license in all new files
- added some imports from `__future__` (needed for python3 transition)
- added some test code
- ran pylint on all new files (except tinynumpy.py)
    - at some point I would like you to run pylint on test_tinydnarray.py again
    - it is scoring a little low now
- ran 2to3 conversions

All tests are passing in:
- python2 with and without numpy
- python3 with and without numpy

I am closing this issue. Good job! Eric.

**Comment by eayoungs** [comment]
Great! Will do, Santosh. Thanks for all the help and encouragement.

**Comment by eayoungs** [comment]
Santosh,
  Can you try installing tinynumpy from pip again? I believe I've got it working now.

Cheers

Eric

**Comment by santoshphilip** [comment]
Yup. 
**pip install tinynumpy** is working fine.
This will ease the 2to3 transiton

**Comment by santoshphilip** [comment]
Can you make changes in your branch so that 
- tinynumpy is in requirements.txt
- remove tinynumpy from the eppy codebase 

Let me know when all is well and I'll merge into *develop* branch

**Comment by eayoungs** [comment]
Sure; I've got those changes mostly completed on the *eayoungs* branch. Just want to verify some things. I will probably be able to complete it this week.

**Comment by eayoungs** [comment]
Hey, sorry for the delay. I got caught up in work last week. I ran into some issues with tinynumpy again. It's not importing into eppy, even though it appears to be installed in my virtualenv. I'll have to dig into PyPI again.
Every time I google issues with PyPI, I run into blog posts that wonder why it's so difficult to get packages working with PyPI, so at least I'm not alone!

Eric

**Comment by santoshphilip** [comment]
for some reason, I did not see the second half of your message. Will test and respond

**Comment by santoshphilip** [comment]
I am still not getting an install for tinynumpy

    (eppy_nonumpy)Santoshs-Air:~ santoshphilip$ pip install tinynumpy
    Downloading/unpacking tinynumpy
      Could not find a version that satisfies the requirement tinynumpy (from versions: 1.2.0.dev3)
    Cleaning up...
    No distributions matching the version for tinynumpy
    Storing debug log for failure in /Users/santoshphilip/.pip/pip.log

shall I try to set up a temporary pipy account for tinynumpy (I'll call it tinynumpy_test or something). Just to see if I can get it to work.

**Comment by eayoungs** [comment]
Ugh! I thought I tested it on my Ubuntu machine, too. Let me check into it. This week is a little tight but I should be able to get to it in the next day or two.
I apologize for all the back & forth but I think it's key that I get this right, especially since my next task is to get the Python3 package working.

Thanks for staying in touch about this issue!

Eric

**Comment by eayoungs** [comment]
Santosh,
Here's what I get on my Ubuntu machine:

*(test_tnp)eayoungs@pinky:~$ pip install tinynumpy
Collecting tinynumpy
  Downloading tinynumpy-1.2.0.dev3.tar.gz
Building wheels for collected packages: tinynumpy
  Running setup.py bdist_wheel for tinynumpy
  Stored in directory: /home/eayoungs/.cache/pip/wheels/9e/93/24/ade532b9bd77231fb222c4c1f6e62947014e568429bd76fcf3
Successfully built tinynumpy
Installing collected packages: tinynumpy
Successfully installed tinynumpy-1.2.0.dev3*

All tests pass on Ubuntu with both setups, too. Can you have another contributor test this?

**Comment by santoshphilip** [comment]
Eric, I tracked down the problem.

From my error log:

    https://pypi.python.org/simple/tinynumpy/), version 1.2.0.dev3 is a pre-release (use --pre to allow).
    
It installs if I do:

    pip install --pre tinynumpy
    
For eppy to install it, it should work without the "--pre".
If you are comfortable enough with the release, can you make a release that is not a pre-release ?

**Comment by eayoungs** [comment]
Interesting. I didn't use that flag, but I'll check with @wadetb (tinynumpy) and see if I can get this updated on PyPI. It may be late this week or early next.

**Comment by santoshphilip** [comment]
I am attaching my error log, if it helps.
[pip.txt](https://github.com/santoshphilip/eppy/files/5310/pip.txt)

**Comment by santoshphilip** [comment]
Maybe this is happening because the word "dev" is in the version number.

the DOAP record: tinynumpy-1.2.0.dev3.xml says

    <?xml version="1.0" encoding="UTF-8" ?>
    <rdf:RDF xmlns="http://usefulinc.com/ns/doap#" xmlns:foaf="http://xmlns.com/foaf/0.1/" xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"><Project><name>tinynumpy</name>
    <shortdesc>A lightweight, pure Python, numpy compliant ndarray class</shortdesc>
    <homepage rdf:resource="https://github.com/wadetb/tinynumpy" />
    <maintainer><foaf:Person><foaf:name>Wade Brainerd, Almar Klein</foaf:name>
    <foaf:mbox_sha1sum>2d85bb5e3da728bde551ac136e647bd44cd904fa</foaf:mbox_sha1sum></foaf:Person></maintainer>
    <release><Version><revision>1.2.0.dev3</revision></Version></release>
    </Project></rdf:RDF>

try without the word "dev" in the version number. 
It might fix the issue

**Comment by eayoungs** [comment]
Sorry for the delay; I'll give it a try today.

**Comment by eayoungs** [comment]
Give it a try now; I've removed the "*dev\d*" from the package name.

**Comment by santoshphilip** [comment]
Yes ! It works

**Comment by santoshphilip** [comment]
in branch `i30_pythongeometry` all tests pass
- with `tinynumpy` installed and
- without `numpy` installed

I have lost track of what to merge, since this is an old issue. My recollection is that this was already merged into branch `develop`. I'll try to spend some time update myself on the state of this issue

**Comment by eayoungs** [comment]
Great! I think I've got all the changes in the *eayoungs* branch. If the current state of that branch has been merged, then we're likely up to date.

Now I can move on to getting the Python 3 version of eppy available on PyPI

**Comment by santoshphilip** [comment]
Eric,
I was looking thru branch `eayoungs` and some thoughts:
- requirements.txt should have tinynumpy in it
- tinnumpy should be removed from eppy (and any associated tests for tinynumpy)
- the import statements for tinynumpy will need an update (or pytest will fail)
- after you do this, all pytest should pass 

I tested the merge and I am not seeing conflicts arising from doing the merge into develop

**Comment by eayoungs** [comment]
Santosh,
I had made most of those changes previously but somehow stashed the changes when I undid the changes to the develop branch.
Anyway, I reapplied the changes and added *tinynumpy>=1.2.0* to *requirements.txt*. All tests still pass with and without numpy. All commits have been pushed  to *eayoungs* branch should now be available on Github.

Cheers

Eric

**Comment by santoshphilip** [comment]
Thanks. I'll test and then merge into develop.
will close this issue once that is done

**Comment by santoshphilip** [comment]
Some import statements need to be updated.
I have tinynumpy installed:

    (tinynumpy)Santoshs-MacBook-Air:eppy santoshphilip$ python 
    Python 2.7.6 (default, Apr  9 2014, 11:48:52) 
    [GCC 4.2.1 Compatible Apple LLVM 5.1 (clang-503.0.38)] on darwin
    Type "help", "copyright", "credits" or "license" for more information.
    >>> import tinynumpy
    >>> 

py.test gives the following errors

    (tinynumpy)Santoshs-MacBook-Air:eppy santoshphilip$ py.test
    ============================= test session starts ==============================
    platform darwin -- Python 2.7.6, pytest-2.8.2, py-1.4.30, pluggy-0.3.1
    rootdir: /Users/santoshphilip/Documents/coolshadow/eppy, inifile: setup.cfg
    collected 12 items / 12 errors 

    eppy/tests/test_bunchhelpers.py ....
    eppy/tests/test_iddgaps.py .
    eppy/tests/test_readhtml.py ......
    eppy/tests/EPlusInterfaceFunctions_tests/test_parse_idd.py .

    ==================================== ERRORS ====================================
    ___________________ ERROR collecting eppy/tests/test_IDF.py ____________________
    eppy/tests/test_IDF.py:18: in <module>
        from eppy.modeleditor import IDF0
    eppy/modeleditor.py:17: in <module>
        from eppy.idfreader import idfreader1
    eppy/idfreader.py:20: in <module>
        import eppy.function_helpers as fh
    eppy/function_helpers.py:17: in <module>
        from eppy.geometry import surface as g_surface
    eppy/geometry/surface.py:27: in <module>
        from . import tinynumpy as np
    E   ImportError: cannot import name tinynumpy
    ______________ ERROR collecting eppy/tests/test_bunch_subclass.py ______________
    eppy/tests/test_bunch_subclass.py:682: in <module>
        import eppy.idfreader as idfreader
    eppy/idfreader.py:20: in <module>
        import eppy.function_helpers as fh
    eppy/function_helpers.py:17: in <module>
        from eppy.geometry import surface as g_surface
    eppy/geometry/surface.py:27: in <module>
        from . import tinynumpy as np
    E   ImportError: cannot import name tinynumpy
    _________________ ERROR collecting eppy/tests/test_examples.py _________________
    eppy/tests/test_examples.py:15: in <module>
        from eppy.idfreader import idfreader
    eppy/idfreader.py:20: in <module>
        import eppy.function_helpers as fh
    eppy/function_helpers.py:17: in <module>
        from eppy.geometry import surface as g_surface
    eppy/geometry/surface.py:27: in <module>
        from . import tinynumpy as np
    E   ImportError: cannot import name tinynumpy
    _______________ ERROR collecting eppy/tests/test_hvacbuilder.py ________________
    eppy/tests/test_hvacbuilder.py:15: in <module>
        import eppy.hvacbuilder as hvacbuilder
    eppy/hvacbuilder.py:17: in <module>
        import eppy.modeleditor as modeleditor
    eppy/modeleditor.py:17: in <module>
        from eppy.idfreader import idfreader1
    eppy/idfreader.py:20: in <module>
        import eppy.function_helpers as fh
    eppy/function_helpers.py:17: in <module>
        from eppy.geometry import surface as g_surface
    eppy/geometry/surface.py:27: in <module>
        from . import tinynumpy as np
    E   ImportError: cannot import name tinynumpy
    ________________ ERROR collecting eppy/tests/test_idfreader.py _________________
    eppy/tests/test_idfreader.py:15: in <module>
        import eppy.idfreader as idfreader
    eppy/idfreader.py:20: in <module>
        import eppy.function_helpers as fh
    eppy/function_helpers.py:17: in <module>
        from eppy.geometry import surface as g_surface
    eppy/geometry/surface.py:27: in <module>
        from . import tinynumpy as np
    E   ImportError: cannot import name tinynumpy
    _______________ ERROR collecting eppy/tests/test_modeleditor.py ________________
    eppy/tests/test_modeleditor.py:18: in <module>
        from eppy.modeleditor import IDF
    eppy/modeleditor.py:17: in <module>
        from eppy.idfreader import idfreader1
    eppy/idfreader.py:20: in <module>
        import eppy.function_helpers as fh
    eppy/function_helpers.py:17: in <module>
        from eppy.geometry import surface as g_surface
    eppy/geometry/surface.py:27: in <module>
        from . import tinynumpy as np
    E   ImportError: cannot import name tinynumpy
    ________________ ERROR collecting eppy/tests/test_simpleread.py ________________
    eppy/tests/test_simpleread.py:14: in <module>
        import eppy.simpleread as simpleread
    eppy/simpleread.py:14: in <module>
        import eppy.modeleditor as modeleditor
    eppy/modeleditor.py:17: in <module>
        from eppy.idfreader import idfreader1
    eppy/idfreader.py:20: in <module>
        import eppy.function_helpers as fh
    eppy/function_helpers.py:17: in <module>
        from eppy.geometry import surface as g_surface
    eppy/geometry/surface.py:27: in <module>
        from . import tinynumpy as np
    E   ImportError: cannot import name tinynumpy
    _________ ERROR collecting eppy/tests/geometry_tests/test_area_zone.py _________
    eppy/tests/geometry_tests/test_area_zone.py:12: in <module>
        import eppy.geometry.area_zone as area_zone
    eppy/geometry/area_zone.py:18: in <module>
        import eppy.geometry.surface as surface
    eppy/geometry/surface.py:27: in <module>
        from . import tinynumpy as np
    E   ImportError: cannot import name tinynumpy
    __________ ERROR collecting eppy/tests/geometry_tests/test_surface.py __________
    eppy/tests/geometry_tests/test_surface.py:12: in <module>
        import eppy.geometry.surface as surface
    eppy/geometry/surface.py:27: in <module>
        from . import tinynumpy as np
    E   ImportError: cannot import name tinynumpy
    ________ ERROR collecting eppy/tests/geometry_tests/test_volume_zone.py ________
    eppy/tests/geometry_tests/test_volume_zone.py:17: in <module>
        import eppy.geometry.volume_zone as volume_zone
    eppy/geometry/volume_zone.py:25: in <module>
        from eppy.geometry.tinynumpy import dot as dot
    E   ImportError: No module named tinynumpy
    _______ ERROR collecting eppy/tests/tinynumpy_tests/test_tinyndarray.py ________
    eppy/tests/tinynumpy_tests/test_tinyndarray.py:13: in <module>
        import eppy.geometry.tinynumpy as tinynumpy
    E   ImportError: No module named tinynumpy
    ________ ERROR collecting eppy/tests/tinynumpy_tests/test_tinynumpy.py _________
    eppy/tests/tinynumpy_tests/test_tinynumpy.py:16: in <module>
        import eppy.geometry.tinynumpy as tnp
    E   ImportError: No module named tinynumpy
    ===================== 12 passed, 12 error in 0.87 seconds ======================
    (tinynumpy)Santoshs-MacBook-Air:eppy santoshphilip$

**Comment by eayoungs** [comment]
That's odd. First off, there shouldn't be any *tinynumpy* tests in the project anymore.

**Comment by eayoungs** [comment]
Santosh,
I've applied the changes to import directly from the pip installed module, but I'm getting an error related to the linalg module in the tinynumpy project. It is a separate file to mimic the structure of numpy and maximize the reuse of code with either numpy or tinynumpy.
Unfortunately I seem to have named the module tinylinalg instead of linalg, which is causing an error of it's own. I will have to change this in the *tinynumpy* project and revise the project on PyPI.

**Comment by santoshphilip** [comment]
give ma a heads up when you are done with that. 
I'll run the tests again.

I am traveling most of December and may be out of contact. We should get some momentum on the python3 stuff before that. (completing the tinynumpy stuff is a precursor to that)
I'll put together some notes in the developer documentation about the present state of python3 version

**Comment by eayoungs** [comment]
I've been thinking about it and it should be just a matter of adjusting the import statement in *eppy*, hopefully without modifying *tinynumpy*
Well, that will give me some motivation to wrap this up and get started on the Python 3 issues by the end of the month. I've been distracted by work, but the good news is I've been using Python & R at work so my programming skills should have at least held, rather than degrading. ;-)

Eric

**Comment by eayoungs** [comment]
OK. I just made the changes to a few import statements and all tests path in both virtual environments (with numpy only, and with tinynumpy only).
Let's try this one more time!

Happy Thanksgiving!

**Comment by santoshphilip** [comment]
Will test it. Then we will be ready for the python3 transition. 

Santosh.

**Comment by santoshphilip** [comment]
Yessss !!! 

- tests pass with numpy
- test pass without numpy

I'll take steps to make this part of the develop branch

**Comment by santoshphilip** [comment]
Eric,
    I was reviewing you code changes and I see some extra files. They are listed below:

- eppy/geometry/plan.md 
- eppy/redict_subclass.py
- eppy/tests/test_cons_uval.py
- eppy/tests/test_redict_subclass.py
- eppy/useful_scripts/InputSim/ASHRAE90.1_SchoolPrimary_STD2007_Salem.idf
- eppy/useful_scripts/InputSim/90.1-07_PrimSch.idf
- eppy/useful_scripts/InputSim/ASHRAE90.1_SchoolPrimary_STD2007_Salem.idf
- eppy/useful_scripts/InputSim/eppy-useful_scripts-test.ipynb
- eppy/useful_scripts/cons_uval.py
- eppy/useful_scripts/internal_loads.py

looks like most of them are from scripting you were doing.
Are any of them useful to keep in eppy? (cons_uval.py looked interesting)
it would be nice to have:

    surface.uvalue() # based on cons_uval.py

**Comment by eayoungs** [comment]
Great! So glad to hear it.

**Comment by eayoungs** [comment]
Yes. I was so excited, and it was Thanksgiving, so I forgot to clean things
up before pushing.
I believe that was the intention of the script.  I'll take a look and see
what it needs.
I'm out of town until tomorrow, but I'll check in when I get back to town.

**Comment by santoshphilip** [comment]
This is fully functional and released with the latest version
